file.go

```diff
@@ -1,20 +1,135 @@
 package s3fs
 
 import (
+	"errors"
+	"fmt"
 	"io"
 	"io/fs"
+	"io/ioutil"
 	"path"
 	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
 )
 
 var (
 	_ fs.File     = (*file)(nil)
 	_ fs.FileInfo = (*fileInfo)(nil)
+	_ io.Seeker   = (*file)(nil)
 )
 
 type file struct {
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
 	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	stat   func() (fs.FileInfo, error)
+	offset int64
+}
+
+func openFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
+	out, err := cl.GetObject(&s3.GetObjectInput{
+		Key:    &name,
+		Bucket: &bucket,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	statFunc := func() (fs.FileInfo, error) {
+		return stat(cl, bucket, name)
+	}
+
+	if out.ContentLength != nil && out.LastModified != nil {
+		// if we got all the information from GetObjectOutput
+		// then we can cache fileinfo instead of making
+		// another call in case Stat is called.
+		statFunc = func() (fs.FileInfo, error) {
+			return &fileInfo{
+				name:    path.Base(name),
+				size:    *out.ContentLength,
+				modTime: *out.LastModified,
+			}, nil
+		}
+	}
+
+	return &file{
+		cl:         cl,
+		bucket:     bucket,
+		name:       name,
+		ReadCloser: out.Body,
+		stat:       statFunc,
+		offset:     0,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.Read(p)
```

Comment:
bug: this will cause stackoverflow

```go
f.ReadCloser.Read(p)
```

Reply:
Maybe you can also double check the tests. They should find such things.

Reply:
Tests are fine, they panic